### PR TITLE
Part 3: Lower more array expressions.

### PR DESCRIPTION
**Pull request by schweitzpgi**
- add more implementations
- add fatal error handler
- misc. refactoring of expression lowering

### CODE REVIEW

File: flang/lib/Lower/ConvertExpr.cpp [hunk 1]
```diff
@@ -387,28 +368,29 @@ class ScalarExprLowering {
   template <int KIND>
   fir::ExtendedValue
   genval(const Fortran::evaluate::ComplexComponent<KIND> &part) {
-    auto lhs = genunbox(part.left());
-    assert(lhs && "unexpected array expression complex component");
-    return extractComplexPart(lhs, part.isImaginaryPart);
+    return extractComplexPart(genunbox(part.left()), part.isImaginaryPart);
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:371]: Should `genunbox` crash if this is called on something that is not an unboxed value ?
I understand that now that array have a different path, these asserts/TODO should not be hit, but I am a bit worried that we may end up in weird segfaults with compiler bugs. In non debug mode, working on empty `mlir::value` does not crash nicely.

**Review by jeanPerier — APPROVED** [review]
Looks good to me. Could you add a few lit tests of array max/pow/character compare... I think it may be worth it especially for characters that proved quite fragile so far.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 1 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:371]: It does.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 2]
```diff
@@ -505,33 +474,47 @@ class ScalarExprLowering {
...
-    if (!lhs || !rhs)
-      TODO("Array expression extremum");
-    llvm::SmallVector<mlir::Value, 2> operands{lhs, rhs};
-    if (op.ordering == Fortran::evaluate::Ordering::Greater)
-      return Fortran::lower::genMax(builder, getLoc(), operands);
-    return Fortran::lower::genMin(builder, getLoc(), operands);
+    switch (op.ordering) {
+    case Fortran::evaluate::Ordering::Greater:
+      return Fortran::lower::genMax(builder, getLoc(),
+                                    llvm::ArrayRef<mlir::Value>{lhs, rhs});
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:480]: `ValueRange` is the new hotness. It looks like the intrinsics are wired for `ArrayRef` everywhere, so I left all that.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 3]
```diff
@@ -505,33 +474,47 @@ class ScalarExprLowering {
...
+    case Fortran::evaluate::Ordering::Less:
+      return Fortran::lower::genMin(builder, getLoc(),
+                                    llvm::ArrayRef<mlir::Value>{lhs, rhs});
+    case Fortran::evaluate::Ordering::Equal:
+      TODO("");
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:485]: I think it is an error, but I did not investigate it. Therefore I left it as a TODO -- as in someone should make sure.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 4]
```diff
@@ -856,20 +816,32 @@ class ScalarExprLowering {
...
-    TODO("");
+    fir::emitFatalError(getLoc(), "array constructor has no rank");
   }
 
-  fir::ExtendedValue gen(const Fortran::evaluate::ComplexPart &) {
-    TODO("complex part");
-  }
-  fir::ExtendedValue genval(const Fortran::evaluate::ComplexPart &) {
-    TODO("complex part");
+  fir::ExtendedValue gen(const Fortran::evaluate::ComplexPart &x) {
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:828]: Yes. I tested this.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 5]
```diff
@@ -1793,7 +1755,8 @@ class ArrayExprLowering {
   template <typename A, typename = std::enable_if_t<Fortran::common::HasMember<
                             A, Fortran::evaluate::TypelessExpression>>>
   CC genarr(const A &x) {
-    // FIXME: replace this with a report_fatal_error if no longer called
+    // Ev::ProcedureDesignator and Ev::ProcedureRef may yield a non-zero rank,
+    // but treat them as scalar values.
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:1759]: It doesn't seem terribly worthwhile to me to add this check. The other cases have ranks that are hard-coded to 0. On the other hand, these cases are _not_ required to be zero, so they have to be handled.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 6]
```diff
@@ -1906,29 +1876,66 @@ class ArrayExprLowering {
...
+      };
+    case Fortran::evaluate::Ordering::Less:
+      return [=](IterSpace iters) -> ExtValue {
+        auto lhs = fir::getBase(lf(iters));
+        auto rhs = fir::getBase(rf(iters));
+        return Fortran::lower::genMin(builder, loc,
+                                      llvm::ArrayRef<mlir::Value>{lhs, rhs});
+      };
+    case Fortran::evaluate::Ordering::Equal:
+      TODO("");
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:1911]: I was kind of surprised this case even came up. Still it needs to be investigated.

**Comment by schweitzpgi** [comment]
Yes. I plan to add some more tests.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeanPerier) [reply flang/lib/Lower/ConvertExpr.cpp:1911]: Ok, I think it stems of the usage of `Fortran::evaluate::Ordering` enum that was not specially made for `evaluate::Extremum`, so not all cases of the enum makes sense there.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 1 above)
Comment (jeanPerier) [reply flang/lib/Lower/ConvertExpr.cpp:371]: Thanks, I saw your addition of `fir::emitFatalError` inside it now. Sorry for bothering you for nothing about it !

File: flang/lib/Lower/ConvertExpr.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:1911]: I'll turn these into errors. I don't see where this is a legit option either.

**Pull request merged**